cache/azblob/azblob.go

```diff
@@ -0,0 +1,299 @@
+		if err != nil {
+			e := cache.ErrSettingToCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+
+		// read test file
+		byt, hit, err := azCache.Get(&key)
+		if err != nil {
+			e := cache.ErrGettingFromCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+		if !hit {
+			//return an error?
+			panic("no hit on writable cache")
+		}
+
+		// test response of writable cache
+		if string(byt) != testMsg {
+			e := cache.ErrGettingFromCache{
+				Err:       fmt.Errorf("incrrect test response %s != %s", string(byt), testMsg),
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+
+		err = azCache.Purge(&key)
+		if err != nil {
+			e := cache.ErrPurgingCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+	}
+
+	return &azCache, nil
+}
+
+type Cache struct {
+	Basepath  string
+	MaxZoom   uint
+	ReadOnly  bool
+	Container azblob.ContainerURL
+}
+
+func roundUp512(n int) int32 {
+	return int32(n + (512 - n%512))
+}
+
+func (azb *Cache) Set(key *cache.Key, val []byte) error {
+	if key.Z > azb.MaxZoom || azb.ReadOnly {
+		return nil
+	}
+
+	blob := azb.makeBlob(key).ToPageBlobURL()
+	ctx := context.Background()
+
+	httpHeaders := azblob.BlobHTTPHeaders{
+		ContentType: "application/x-protobuf",
+	}
+
+	// must send things in multiples of 512 byte pages
+	msgLen := len(val)
+	blobLen := roundUp512(msgLen + 8)
+
+	// allocate blob
+	blobSlice := make([]byte, blobLen)
+	// encode the length of the blob
+	binary.BigEndian.PutUint64(blobSlice[:8], uint64(msgLen))
+	copy(blobSlice[8:], val)
+
+	_, err := blob.Create(ctx, int64(blobLen), 0, httpHeaders, azblob.Metadata{}, azblob.BlobAccessConditions{})
+	if err != nil {
+		println("\n\n ERR CREATE \n\n")
+		return err
+	}
+
+	pageRange := azblob.PageRange{
+		Start: 0,
+		End:   blobLen - 1,
+	}
+
+	_, err = blob.PutPages(ctx, pageRange, bytes.NewReader(blobSlice), azblob.BlobAccessConditions{})
+	if err != nil {
+		println("\n\n ERR PUT \n\n")
+		return err
+	}
+
+	return nil
+}
+
+func (azb *Cache) Get(key *cache.Key) ([]byte, bool, error) {
+	if key.Z > azb.MaxZoom {
+		return nil, false, nil
+	}
+
+	blob := azb.makeBlob(key)
+	ctx := context.Background()
+
+	res, err := blob.GetBlob(ctx, azblob.BlobRange{}, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		// check if 404
+		resErr, ok := err.(azblob.ResponseError)
+		if ok {
+			if resErr.Response().StatusCode == http.StatusNotFound {
+				return nil, false, nil
+			}
+		}
+
+		return nil, false, err
+	}
+	defer res.Body().Close()
+
+	blobSlice, err := ioutil.ReadAll(res.Body())
+	if err != nil {
+		return nil, false, err
+	}
+
+	// get the encoded message length
+	msgLen := binary.BigEndian.Uint64(blobSlice[:8])
+
+	// check for out of bounds
+	if msgLen > uint64(len(blobSlice)) {
+		return nil, false, fmt.Errorf("azblob: length section does not match message length")
+	}
+
+	return blobSlice[8:][:msgLen], true, nil
```

Comment:
rather than using the magic number `8` throughout, create a constant called `FileHeader` or something similar that can be used throughout.

Reply:
good idea